### PR TITLE
GG-27176

### PR DESCRIPTION
**Pull request by ygerzhedovich**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/jdbc/JdbcRequestHandler.java [hunk 1]
```diff
@@ -506,6 +507,10 @@ private JdbcResponse processBulkLoadFileBatch(JdbcBulkLoadBatchRequest req) {
         // Write node id.
         if (protocolVer.compareTo(VER_2_8_0) >= 0)
             writer.writeUuid(connCtx.kernalContext().localNodeId());
+
+        // Write all features supported by the node.
+        if (protocolVer.compareTo(VER_2_8_2) >= 0)
+            writer.writeByteArray(JdbcThinFeature.allFeatures());
```
Comment (isapego) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/jdbc/JdbcRequestHandler.java]: Matching should happen in a single place, not on both client and server. Or we can face de-synchronisation as we update our clients and servers.

File: modules/core/src/main/java/org/apache/ignite/internal/AbstractThinProtocolFeature.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    /**
+     * @param features Features set.
+     * @return Byte array representing all supported features by current node.
+     * @param <E> Features type (JDBC, thin cli, ODBC)
+     */
+    static <E extends Enum<E> & AbstractThinProtocolFeature> byte[] features(E[] features) {
+        final BitSet set = new BitSet();
+
+        for (AbstractThinProtocolFeature f : features) {
+            final int featureBit = f.bitIdx() + (f.byteIdx() << 3);
```
[the remark this answers is not in the document]
Comment (tledkov-gridgain) [reply modules/core/src/main/java/org/apache/ignite/internal/AbstractThinProtocolFeature.java]: Fixed.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/jdbc/JdbcBinaryContext.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+package org.apache.ignite.internal.processors.odbc.jdbc;
+
+import java.util.EnumSet;
+import org.apache.ignite.internal.processors.odbc.ClientListenerProtocolVersion;
+
+/**
+ * Binary context for JDBC protocol. Holds protocol settings and state.
+ */
+public class JdbcBinaryContext {
```
[the remark this answers is not in the document]
Comment (tledkov-gridgain) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/jdbc/JdbcBinaryContext.java]: Fixed

File: modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/jdbc/JdbcRequestHandler.java (hunk 1 above)
Comment (tledkov-gridgain) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/jdbc/JdbcRequestHandler.java]: Fixed.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/jdbc/JdbcBatchExecuteRequest.java [hunk 4]
```diff
@@ -133,31 +132,33 @@ public boolean isLastStreamBatch() {
...
         }
         else
             writer.writeInt(0);
 
-        if (ver.compareTo(VER_2_4_0) >= 0)
+        if (binCtx.version().compareTo(VER_2_4_0) >= 0)
             writer.writeBoolean(lastStreamBatch);
 
-        if (ver.compareTo(VER_2_7_0) >= 0)
+        if (binCtx.version().compareTo(VER_2_7_0) >= 0)
```
[the remark this answers is not in the document]
Comment (tledkov-gridgain) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/jdbc/JdbcBatchExecuteRequest.java]: fixed

File: modules/core/src/main/java/org/apache/ignite/internal/jdbc/thin/JdbcThinTcpIo.java [hunk 5]
```diff
@@ -138,6 +144,9 @@
     /** Current protocol version used to connection to Ignite. */
     private final ClientListenerProtocolVersion srvProtoVer;
 
+    /** Features set are supported by the current protocol. */
```
Comment (ptupitsyn) [inline modules/core/src/main/java/org/apache/ignite/internal/jdbc/thin/JdbcThinTcpIo.java]: Comment looks misleading

File: modules/core/src/main/java/org/apache/ignite/internal/ThinProtocolFeature.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        return set.toByteArray();
+    }
+
+    /**
+     * Create EnumSet of supported features encoded by the bytes array.
+     *
+     * @param <E> Features type (JDBC, thin cli, ODBC)
+     * @param in Byte array representing all supported features.
+     */
+    static <E extends Enum<E> & ThinProtocolFeature> EnumSet<E> enumSet(byte [] in, E[] values) {
```
Comment (korlov42) [inline modules/core/src/main/java/org/apache/ignite/internal/ThinProtocolFeature.java]: I would change second parameter to `Class<E> cls`. It should make API less error prone.

```
static <E extends Enum<E> & ThinProtocolFeature> EnumSet<E> enumSet(byte [] in, Class<E> cls) {
    ...
    EnumSet<E> set = EnumSet.noneOf(cls);
    ...
    for (E e : cls.getEnumConstants()) {
    ...
```

File: modules/core/src/main/java/org/apache/ignite/internal/jdbc/thin/JdbcThinTcpIo.java [hunk 7]
```diff
@@ -138,6 +144,9 @@
...
+    private JdbcProtocolContext binCtx;
```
Comment (korlov42) [inline modules/core/src/main/java/org/apache/ignite/internal/jdbc/thin/JdbcThinTcpIo.java]: Please rename a variable too

File: modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/jdbc/JdbcMessageParser.java [hunk 8]
```diff
@@ -34,20 +33,20 @@
     /** Kernal context. */
     private final GridKernalContext ctx;
 
-    /** Client protocol version. */
-    private final ClientListenerProtocolVersion ver;
+    /** Binary context. */
```
Comment (korlov42) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/jdbc/JdbcMessageParser.java]: Misleading comment here

File: modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/jdbc/JdbcMessageParser.java [hunk 9]
```diff
@@ -34,20 +33,20 @@
...
+    private final JdbcProtocolContext protoCtx;
 
     /** Initial output stream capacity. */
     protected static final int INIT_CAP = 1024;
 
     /**
      * @param ctx Context.
-     * @param ver Client protocol version.
+     * @param protoCtx Binary context.
```
Comment (korlov42) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/jdbc/JdbcMessageParser.java]: and here

File: modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/jdbc/JdbcProtocolContext.java [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    public boolean isAffinityAwarenessSupported() {
+        return ver.compareTo(VER_2_8_0) >= 0;
+    }
+
+    /**
+     * @return Supported features.
+     */
+    public EnumSet<JdbcThinFeature> features() {
+        return features;
+    }}
```
Comment (korlov42) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/jdbc/JdbcProtocolContext.java]: please move last bracket at the new line

File: modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/jdbc/JdbcThinFeature.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+/**
+ * Defines supported features for JDBC thin client.
+ */
+public enum JdbcThinFeature implements ThinProtocolFeature {
+    RESERVED(0);
+
+    /** */
+    private static final EnumSet<JdbcThinFeature> ALL_FEATURES_AS_ENUM_SET = EnumSet.allOf(
+        JdbcThinFeature.values()[0].getDeclaringClass());
```
Comment (pavlukhin) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/jdbc/JdbcThinFeature.java]: Will `EnumSet.allOf(JdbcThinFeature.class)` work here?

File: modules/core/src/main/java/org/apache/ignite/internal/ThinProtocolFeature.java [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+ * The base feature class.
+ */
+public interface ThinProtocolFeature {
+    /**
+     * @return Feature ID.
+     */
+    int featureId();
+
+    /**
+     * @return Feature's name.
```
Comment (pavlukhin) [inline modules/core/src/main/java/org/apache/ignite/internal/ThinProtocolFeature.java]: Spellchecker: `Feature name`

File: modules/core/src/main/java/org/apache/ignite/internal/ThinProtocolFeature.java [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+     */
+    String name();
+
+    /**
+     * @param features Features set.
```
Comment (pavlukhin) [inline modules/core/src/main/java/org/apache/ignite/internal/ThinProtocolFeature.java]: Spellchecker: `Feature set`. Here and below.

File: modules/core/src/main/java/org/apache/ignite/internal/ThinProtocolFeature.java [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+     * @return Byte array representing all supported features by current node.
+     * @param <E> Features type (JDBC, thin cli, ODBC)
```
Comment (pavlukhin) [inline modules/core/src/main/java/org/apache/ignite/internal/ThinProtocolFeature.java]: Quite unusual to see `@param` after `@return`. Here and below.

Comment (pavlukhin) [inline modules/core/src/main/java/org/apache/ignite/internal/ThinProtocolFeature.java]: Spellchecker: `Feature type`. Here and below.

File: modules/core/src/main/java/org/apache/ignite/internal/ThinProtocolFeature.java [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+            set.set(f.featureId());
+
+        return set.toByteArray();
+    }
+
+    /**
+     * Create EnumSet of supported features encoded by the bytes array.
+     *
+     * @param <E> Features type (JDBC, thin cli, ODBC)
+     * @param in Byte array representing all supported features.
```
Comment (pavlukhin) [inline modules/core/src/main/java/org/apache/ignite/internal/ThinProtocolFeature.java]: `@param enumCls` is missing.

File: modules/core/src/main/java/org/apache/ignite/internal/ThinProtocolFeature.java [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+     */
+    static <E extends Enum<E> & ThinProtocolFeature> EnumSet<E> enumSet(byte [] in, Class<E> enumCls) {
+        final BitSet bSet = BitSet.valueOf(in);
```
Comment (pavlukhin) [inline modules/core/src/main/java/org/apache/ignite/internal/ThinProtocolFeature.java]: NPE if `in == null`?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/jdbc/JdbcThinFeature.java [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+
+import java.util.Arrays;
+import java.util.EnumSet;
+import org.apache.ignite.internal.ThinProtocolFeature;
+
+/**
+ * Defines supported features for JDBC thin client.
+ */
+public enum JdbcThinFeature implements ThinProtocolFeature {
+    RESERVED(0);
```
Comment (pavlukhin) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/jdbc/JdbcThinFeature.java]: Is it really reserved? Or just for illustrative purposes so far?

File: modules/core/src/main/java/org/apache/ignite/internal/ThinProtocolFeature.java (hunk 16 above)
Comment (tledkov-gridgain) [reply modules/core/src/main/java/org/apache/ignite/internal/ThinProtocolFeature.java]: fixed

File: modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/jdbc/JdbcUtils.java [hunk 18]
```diff
@@ -33,7 +33,7 @@
      * @param writer Binary writer.
      * @param items Query results items.
      */
-    public static void writeItems(BinaryWriterExImpl writer, List<List<Object>> items) {
+    public static void writeItems(BinaryWriterExImpl writer, List<List<Object>> items, JdbcProtocolContext protoCtx) {
```
Comment (mshonichev) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/jdbc/JdbcUtils.java]: why do we add `protoCtx` here and don't use it? 
also, please add to javadoc

File: modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/jdbc/JdbcUtils.java [hunk 19]
```diff
@@ -50,7 +50,7 @@ public static void writeItems(BinaryWriterExImpl writer, List<List<Object>> item
      * @param reader Binary reader.
      * @return Query results items.
      */
-    public static List<List<Object>> readItems(BinaryReaderExImpl reader) {
+    public static List<List<Object>> readItems(BinaryReaderExImpl reader, JdbcProtocolContext ctx) {
```
Comment (mshonichev) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/jdbc/JdbcUtils.java]: same as above

File: modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/jdbc/JdbcThinFeature.java [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+     * @return Set of supported features.
+     */
+    public static EnumSet<JdbcThinFeature> enumSet(byte[] bytes) {
+        return ThinProtocolFeature.enumSet(bytes, JdbcThinFeature.class);
+    }
+
+    /**
+     * @return Byte array representing all supported features by current node.
+     */
+    public static byte[] allFeaturesAsBytes() {
```
Comment (mshonichev) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/jdbc/JdbcThinFeature.java]: I don't understand, why does `JdbcThinFeature` provides custom `allFeaturesAsBytes()` when `ThinProtocolFeature` has `featuresAsBytes` that seems to do the same as method description suggests. What's the purpose (e.g. usecase) of `featuresAsBytes` then?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/jdbc/JdbcThinFeature.java [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+package org.apache.ignite.internal.processors.odbc.jdbc;
+
+import java.util.Arrays;
+import java.util.EnumSet;
+import org.apache.ignite.internal.ThinProtocolFeature;
+
+/**
+ * Defines supported features for JDBC thin client.
+ */
+public enum JdbcThinFeature implements ThinProtocolFeature {
```
Comment (mshonichev) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/jdbc/JdbcThinFeature.java]: Would all other thin protocols have their own `XXXThinFeature` subclasses? What about .NET features?

Comment (tledkov-gridgain) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/jdbc/JdbcThinFeature.java]: There is not .net features.
ODBC and thin client will be implemented on separate tickets.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/jdbc/JdbcThinFeature.java (hunk 20 above)
Comment (tledkov-gridgain) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/jdbc/JdbcThinFeature.java]: Take a look at the implementation.
`ThinProtocolFeature#featuresAsBytes(E[])` based on the loop.
`JdbcThinFeature#allFeaturesAsBytes` use cached value.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/jdbc/JdbcUtils.java (hunk 19 above)
Comment (tledkov-gridgain) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/jdbc/JdbcUtils.java]: You right. 
This change may be applied when it is needed.
Reverted.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/jdbc/JdbcUtils.java (hunk 18 above)
Comment (tledkov-gridgain) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/jdbc/JdbcUtils.java]: You right.
This change may be applied when it is needed.
Reverted.

**Pull request closed without merging**